src/backends/webgl2/WebGLQuadShader.ts

```diff
@@ -0,0 +1,251 @@
+    }
+
+    this.#program = program;
+
+    // Get uniform locations
+    this.#uViewProjection = gl.getUniformLocation(program, "u_viewProjection");
+    this.#uResolution = gl.getUniformLocation(program, "u_resolution");
+    this.#uTextureArray = gl.getUniformLocation(program, "u_textureArray");
+
+    // Create VAO
+    const vao = gl.createVertexArray();
+    assert(vao, "Failed to create WebGL VAO");
+    this.#vao = vao;
+
+    // Create instance buffer
+    const instanceBuffer = gl.createBuffer();
+    assert(instanceBuffer, "Failed to create WebGL instance buffer");
+    this.#instanceBuffer = instanceBuffer;
+
+    // Allocate CPU buffer
+    this.#cpuBuffer = new Float32Array(instanceCount * INSTANCE_FLOATS);
+
+    // Set up VAO
+    gl.bindVertexArray(vao);
+    gl.bindBuffer(gl.ARRAY_BUFFER, instanceBuffer);
+    gl.bufferData(
+      gl.ARRAY_BUFFER,
+      instanceCount * INSTANCE_BYTES,
+      gl.DYNAMIC_DRAW,
+    );
+
+    // Set up instance attributes
+    // Each vec4 attribute takes up 16 bytes
+    // model0 at location 0
+    gl.enableVertexAttribArray(0);
+    gl.vertexAttribPointer(0, 4, gl.FLOAT, false, INSTANCE_BYTES, 0);
+    gl.vertexAttribDivisor(0, 1);
+
+    // model1 at location 1
+    gl.enableVertexAttribArray(1);
+    gl.vertexAttribPointer(1, 4, gl.FLOAT, false, INSTANCE_BYTES, 16);
+    gl.vertexAttribDivisor(1, 1);
+
+    // model2 at location 2
+    gl.enableVertexAttribArray(2);
+    gl.vertexAttribPointer(2, 4, gl.FLOAT, false, INSTANCE_BYTES, 32);
+    gl.vertexAttribDivisor(2, 1);
+
+    // tint at location 3
+    gl.enableVertexAttribArray(3);
+    gl.vertexAttribPointer(3, 4, gl.FLOAT, false, INSTANCE_BYTES, 48);
+    gl.vertexAttribDivisor(3, 1);
+
+    // uvOffsetAndScale at location 4
+    gl.enableVertexAttribArray(4);
+    gl.vertexAttribPointer(4, 4, gl.FLOAT, false, INSTANCE_BYTES, 64);
+    gl.vertexAttribDivisor(4, 1);
+
+    // cropOffsetAndScale at location 5
+    gl.enableVertexAttribArray(5);
+    gl.vertexAttribPointer(5, 4, gl.FLOAT, false, INSTANCE_BYTES, 80);
+    gl.vertexAttribDivisor(5, 1);
+
+    // atlasIndex at location 6 (stored as float for simplicity)
+    gl.enableVertexAttribArray(6);
+    gl.vertexAttribPointer(6, 1, gl.FLOAT, false, INSTANCE_BYTES, 96);
+    gl.vertexAttribDivisor(6, 1);
+
+    gl.bindVertexArray(null);
+    gl.bindBuffer(gl.ARRAY_BUFFER, null);
+
+    // Cleanup shaders (they're linked to the program now)
+    gl.deleteShader(vs);
+    gl.deleteShader(fs);
+  }
+
+  startFrame(uniform: EngineUniform): void {
+    this.#instanceIndex = 0;
+
+    const gl = this.#backend.gl;
+    gl.useProgram(this.#program);
+
+    // Set uniforms
+    if (this.#uViewProjection) {
+      // wgpu-matrix mat3 is stored as 12 floats (3 columns × 4 floats with padding)
+      // WebGL uniformMatrix3fv expects 9 floats, so extract the relevant values
+      const m = uniform.viewProjectionMatrix;
+      const mat3x3 = new Float32Array([
```

Comment:
seems like this should be reused between frames, couldn't we update it in updateEngineUniform?